app/src/main/java/com/babylon/wallet/android/presentation/dapp/unauthorized/accountonetime/OneTimeChooseAccountsViewModel.kt

```diff
@@ -91,19 +121,86 @@ class OneTimeChooseAccountsViewModel @Inject constructor(
     private fun isSingleChoice(): Boolean {
         return args.numberOfAccounts == 1 && args.isExactAccountsCount
     }
+
+    fun onContinueClick() {
+        state.value.walletUnauthorizedRequest?.let { request ->
+            setSigningInProgress(true)
+
+            viewModelScope.launch {
+                val selectedAccountEntities = state.value.selectedAccounts().mapNotNull {
+                    getProfileUseCase().activeAccountOnCurrentNetwork(it.address)
+                }.map { it.asProfileEntity() }
+
+                // check if signature is required
+                if (request.oneTimeAccountsRequestItem?.challenge != null) {
+                    collectSignatures(
+                        challenge = request.oneTimeAccountsRequestItem.challenge,
+                        selectedAccountEntities = selectedAccountEntities,
+                        metadata = request.metadata
+                    )
+                } else { // otherwise return the collected accounts without signatures
+                    sendEvent(
+                        OneTimeChooseAccountsEvent.AccountsCollected(
+                            accountsWithSignatures = selectedAccountEntities.associateWith { null }
```

Comment:
Minor observation: it's kind of weird to see a map with nullable values, and even more weird to see the line where all the map values are null.  Perhaps, an improvement would be to reflect what happens exactly through some helper models. Maybe instead of a Map, there might be a sealed interface that would reflect both cases: one would contain all the data (can be the same Map, but with non-nullable signatures), while the other only the selected account entities.

Reply:
yeah... I see your point, and I have another idea. Since it’s clear from the code that signatures are only needed when the request requires authentication—otherwise, just the accounts (hence the null)—how about renaming the parameter to `accountsWithOptionalSignatures`?